pipeline-runner/steps/preparepipeline/internal/context_builder.go

```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		return "", "", err
+	}
+	if err = radixvalidators.GitTagsContainIllegalChars(commitTags); err != nil {
+		return "", "", err
+	}
+	return pipelineTargetCommitHash, commitTags, nil
+}
+
+func (cb *contextBuilder) analyseSourceRepositoryChanges(pipelineInfo *model.PipelineInfo, targetEnvironments []string, pipelineTargetCommitHash string) (bool, []model.EnvironmentToBuild, error) {
+	radixDeploymentCommitHashProvider := commithash.NewProvider(cb.kubeClient, cb.radixClient, pipelineInfo.GetAppName(), targetEnvironments)
+	lastCommitHashesForEnvs, err := radixDeploymentCommitHashProvider.GetLastCommitHashesForEnvironments()
+	if err != nil {
+		return false, nil, err
+	}
+
+	changesFromGitRepository, radixConfigWasChanged, err := git.GetChangesFromGitRepository(pipelineInfo.GetGitWorkspace(),
+		pipelineInfo.GetRadixConfigBranch(),
+		pipelineInfo.GetRadixConfigFileInWorkspace(),
+		pipelineTargetCommitHash,
+		lastCommitHashesForEnvs)
+	if err != nil {
+		return false, nil, err
+	}
+
+	environmentsToBuild := cb.getEnvironmentsToBuild(pipelineInfo, changesFromGitRepository)
+	return radixConfigWasChanged, environmentsToBuild, nil
+}
+
+func (cb *contextBuilder) getEnvironmentsToBuild(pipelineInfo *model.PipelineInfo, changesFromGitRepository map[string][]string) []model.EnvironmentToBuild {
+	var environmentsToBuild []model.EnvironmentToBuild
+	for envName, changedFolders := range changesFromGitRepository {
+		var componentsWithChangedSource []string
+		for _, radixComponent := range pipelineInfo.GetRadixApplication().Spec.Components {
+			if ComponentHasChangedSource(envName, &radixComponent, changedFolders) {
+				componentsWithChangedSource = append(componentsWithChangedSource, radixComponent.GetName())
+			}
+		}
+		for _, radixJobComponent := range pipelineInfo.GetRadixApplication().Spec.Jobs {
+			if ComponentHasChangedSource(envName, &radixJobComponent, changedFolders) {
+				componentsWithChangedSource = append(componentsWithChangedSource, radixJobComponent.GetName())
+			}
+		}
+		environmentsToBuild = append(environmentsToBuild, model.EnvironmentToBuild{
+			Environment: envName,
+			Components:  componentsWithChangedSource,
+		})
+	}
+	return environmentsToBuild
+}
+
+func getGitHash(pipelineInfo *model.PipelineInfo) (string, error) {
+	// getGitHash return git commit to which the user repository should be reset before parsing sub-pipelines.
+	pipelineArgs := pipelineInfo.PipelineArguments
+	pipelineType := pipelineInfo.GetRadixPipelineType()
+	if pipelineType == v1.ApplyConfig {
+		return "", nil
+	}
+
+	if pipelineType == v1.Promote {
+		sourceRdHashFromAnnotation := pipelineInfo.SourceDeploymentGitCommitHash
+		sourceDeploymentGitBranch := pipelineInfo.SourceDeploymentGitBranch
+		if sourceRdHashFromAnnotation != "" {
+			return sourceRdHashFromAnnotation, nil
+		}
+		if sourceDeploymentGitBranch == "" {
+			log.Info().Msg("source deployment has no git metadata, skipping sub-pipelines")
```

Comment:
```suggestion
			log.Info().Msg("Source deployment has no git metadata, skipping sub-pipelines")
```
Capitalize first letter of sentence in log messages

Reply:
Fixed